Dns proxy use original source address and port

jrajahalme opened the pull request:
Make Cilium DNS proxy transparent by using the original source address (and port) in upstream connections. Since clients can issue multiple requests from the same port without waiting for the responses in between, we must support multiple requests in flight at the same time on the client side as well. To this end we share the DNS client between all requests that use the same upstream 5-tuple.

Support for shared clients is imported from a new version of cilium/dns.

```release-note
Cilium DNS proxy now uses the original pod's address as the source address towards the DNS servers.
```

bimmlerd commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -783,6 +789,117 @@ func setSoMarks(fd int, ipFamily ipfamily.IPFamily, secId identity.NumericIdenti
+	var resp Response
+	for {
+		// Try taking the reader lock
+		if c.reader.TryLock() {
```
As the docs say: `Note that while correct uses of TryLock do exist, they are rare, and use of TryLock is often a sign of a deeper problem in a particular use of mutexes.`

I'm not yet convinced that this is such a case. The code is somewhat complicated, and reasoning about it in terms of deadlocks or race conditions thus difficult.

Could you explain the reasoning behind taking this approach a bit more? On what grounds do you think caching the socket is worth the additional complexity here? Or is the intention to make the proxy transparent the intent of the PR, and this complexity is just the fall-out?

jrajahalme replied:
Yes, the intention is to make the DNS proxy transparent to make it play nice with IPsec. As a consequence we must support multiple requests on the same connection (5-tuple), as the DNS clients are doing it (e.g. curl without the `-4` flag does `A` and `AAAA` queries from the same socket back-to-back without waiting for a response).

I first tried an approach where each DNS client is served by a goroutine, but also adds more complexity. While being easier to reason about locking behavior, it adds the overhead of an additional goroutine to each DNS request (minus 5-tuple reuse).

Looking at the `cilium/dns` code I think it is not safe to allow more than one goroutine read from a client's connection as there is state that is not protected against concurrent access. I tried to keep `cilium/dns` changes to a minimum to not expand the scope of this PR too much, so I went with the current approach, where the control flow is almost the same as before, with the addition of passing responses to waiting requestors and a remaining waiter taking over the reading once the current reader gets its own response.

(in fact at first I was just running multiple clients on the same 5-tuple which was possible due to reuseport option, but that resulted in a wrong client reading a response and trashing it for a non-matching ID. This would likely have failed for TCP as well.)

All this being said I'm all open for ideas to make this better :-)

jrajahalme replied:
@bimmlerd It should be noted that each `curl` invocation in a test with toFQDNs policy stresses this concurrency pattern due to the back-to-back `A` and `AAAA` requests I mentioned above.

jrajahalme replied:
Also, the read/write deadlines set on the connection will make the blocking read error out on timeout, so the read loop will terminate on that.

jrajahalme replied:
Also, we do not know about the need for the shared client until the 2nd request comes in while the 1st response is still pending. At that point it is too late to schedule a shared reader (via a goroutine) as the 1st requestor is already blocking on the read. Hence we would need to schedule a goroutine already when the 1st request is being served, which adds per-request overhead.

We could do a more involving redesign for UDP where we would only use one raw UDP socket with per-packet control block for controlling the local address, like we already do for getting the UDP requests in in the first place. In that case we could still need a goroutine per TCP connection, unless we do this kind of "co-operative" reading from the shared client connection.

jrajahalme replied:
For reasoning about locking:
- `c.reader.TryLock` can not block
- there is only one other lock in this pattern, therefore `c.Lock()` can not deadlock, unless there is a path where a goroutine does not eventually call `c.Unlock()`. We do `defer c.Unlock()` so this is highly unlikely
- I had this pattern deadlock on development, though, on write to the reader's own channel. I made the channel buffered for that reason. If there is a path where there are more than one write to any of the `reqs` channels, there could be a deadlock. This is why we delete the channel from the map before writing on the channel, which makes multiple writes to any one channel also impossible

bimmlerd replied:
Did you think about which locks are held outside of the DNS proxy when control flow enters these critical sections? Specifically, are we allowed to hold the DNSProxy lock here, or does that create a potential wait cycle involving Endpoints & ipcache?

bimmlerd reviewed (commented):
I know it's still a draft, but I've briefly looked through and the code seems to have become more complex; thus questions around the approach arose

rolinh commented on go.mod:
```diff
@@ -267,3 +267,5 @@ replace (
 	sigs.k8s.io/controller-tools => github.com/cilium/controller-tools v0.8.0-1
 )
+
+replace github.com/sasha-s/go-deadlock => github.com/jrajahalme/go-deadlock v0.0.0-20231103061022-4ce5b1a047c2
```
What is the reason for the new replace directive? We're trying very hard to get rid of all of them as they create a lot of issues for downstream projects.

jrajahalme replied:
Intended to fix this before undrafting this, but forgot. We now have https://github.com/cilium/go-deadlock, will make this PR use that instead.

jrajahalme replied:
Fixed.

bimmlerd commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -759,16 +765,141 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
+	reqs map[uint16]chan Response // outstanding requests
+
+	// 'reader' mutex is used to serialize reads on 'conn'. It is always taken and released
+	// while holding the lock but the lock can be released and re-acquired while holding
```
nit: `the lock` is a bit confusing for a struct with multiple locks. I read this as refering to the RWMutex on line 810- can we come up with a name for that?

second nit: maybe move the `reader` mutex closer to the `conn` it serialises reads for, and call it `connRLock` or similar

jrajahalme replied:
changed the verbiage to "main lock" and renamed `reader` as `readerLock`. Moved 'conn' to the end right below `readerLock` to signify it is protected by it. Reading from `dns.Client` modifies its state, so an RLock is not suitable for this.

bimmlerd commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -759,16 +765,141 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
+// A Client keeps state for concurrent transactions on the same upstream connection.
+type Client struct {
+	dns.Client
+	refcount int // protected by proxy's lock
```
Does protecting this using the proxy's lock scale? This implies the proxy lock is taken on every DNS request, correct? If so, I don't think that's okay, since we're AFAIK going to great lengths to avoid serialising these, see `getMutexesForResponseIPs` and friends.

jrajahalme replied:
A wider lock is needed to manage the map of the (shared) clients anyway. Logically that lock is on the same scope as the proxy's lock, but maybe I should introduce another lock in `DNSProxy` for this purpose?

bimmlerd replied:
I agree that the lock is necessary for the client cache. I worry that this implies serialising all DNS requests, when they previously were not.

Specifically, it looks to me that we'll be holding the `DNSProxy` lock while creating the socket (syscall, AFAIK), which sounds like it'll regress performance. Do you have a way to compare performance before and after this change?

jrajahalme replied:
I do not have a way to compare performance before and after this change, unfortunately.

I'll see if I can minimize these critical sections in any way. It looks like on close the lock must be held while closing to avoid the case where a new connection with the same 5-tuple is created before the old one is actually closed.

nebril replied:
Can we have a map of locks for the current client keys? This way the proxy-wide lock only protects adding the lock to map (on first DNS request with given key/5-tuple), client creation can be done while holding the client-key-specific lock, which should help with parallelizing DNS client creation.

jrajahalme replied:
Moved Dial to happen under client's own lock instead of the proxy lock, so all related system call overhead is not gone from the proxy's critical section. Unfortunately it looks like the related close must happen under the lock governing the map of clients, otherwise a bind error is possible.

jrajahalme replied:
Now that the syscalls are out of the `NewDNSClient()` this is essentially the same as having a map of locks for the currently in-use keys, IMO.

rolinh reviewed (approved):
Vendor changes lgtm

bimmlerd commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -910,36 +1050,83 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
+			p.sendRefused(scopedLog, w, request)
+			return
+		}
+		if key != "" {
```
nit: I'd prefer `if useOriginalSrc` for consistency

jrajahalme replied:
Fixed, also in the deferred close function.

bimmlerd commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -910,36 +1050,83 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
+		client.refcount--
+		if client.refcount == 0 {
+			// Make client unreachable and close it's connection
+			if key != "" {
```
nit: I'd prefer `if useOriginalSrc` for consistency

jrajahalme replied:
Fixed.

bimmlerd commented on pkg/lock/lock_debug.go:
```diff
@@ -81,6 +91,12 @@ func (i *internalMutex) Lock() {
 
+func (i *internalMutex) TryLock() bool {
+	res := i.Mutex.TryLock()
+	i.Time = time.Now()
```
should this write to `i.Time` if res is `false`?

jrajahalme replied:
Maybe not, but it seems also harmless, as `i.Time` is only used in `Unlock` and in the case `TryLock` failing there is always another write to `i.Time` before the next `Unlock`.

Made the setting of `i.Time` condtional on `res`.

bimmlerd replied:
Ah, yeah that reasoning makes a lot of sense- either way, resolved.

bimmlerd commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -759,16 +765,141 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
+	c.Lock()
+	defer c.Unlock()
+	if _, exists := c.reqs[m.Id]; exists {
+		return nil, 0, fmt.Errorf("Duplicate request: %d", m.Id)
```
nit: Go errors should be lower case
```suggestion
		return nil, 0, fmt.Errorf("duplicate request: %d", m.Id)
```

jrajahalme replied:
fixed, thanks!

bimmlerd commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -759,16 +765,141 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
+	// Send while holding the client lock, as dns.Client is not made to be usable from
+	// concurrent goroutines.
+	start := time.Now()
+	err = c.SendContext(context.Background(), m, c.conn, start)
```
Shouldn't we have a request-scoped context available to use here?

jrajahalme replied:
If we do, it was not possible to use it before. Will investigate.

jrajahalme replied:
There is no request-scoped context. There is this TODO when concurrency limit is in use, in the very beginning of `ServeDNS()`:
```
	if p.ConcurrencyLimit != nil {
		// TODO: Consider plumbing the daemon context here.
		ctx, cancel := context.WithTimeout(context.TODO(), p.ConcurrencyGracePeriod)
		defer cancel()
```

bimmlerd replied:
Yeah fair enough, unfortunate that there isn't one.

bimmlerd commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -759,16 +765,141 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
+		// Unlock for the blocking duration to allow concurrent writes
+		// on the client's connection.
+		c.Unlock()
+		resp = <-ch
```
Why doesn't this block a reader who lost the `TryLock` forever if there is no response for the request ID it is waiting for? Would that leak a goroutine?

jrajahalme replied:
Each reader takes turns on reading for as long as it gets its response, or connection of the `dns.Client` fails. `dns` package sets deadlines on the connection on every read, so those will fail the connection if there is no response. In this regard this "co-operative" reading behavior is no different from a single blocking call to `c.conn.ReadMsg()`.

from Go docs:
```
    // SetReadDeadline sets the deadline for future Read calls.
    // A zero value for t means Read will not time out.
    SetReadDeadline(t time.Time) error

    // SetWriteDeadline sets the deadline for future Write calls.
    // Even if write times out, it may return n > 0, indicating that
    // some of the data was successfully written.
    // A zero value for t means Write will not time out.
    SetWriteDeadline(t time.Time) error
```
From this I read that the deadline will affect all future (read) calls, not just the next one.

Setting of the deadline is here: https://github.com/cilium/dns/blob/master/client.go#L249-L265

It gets configured here: https://github.com/cilium/cilium/blob/362b80ef8548db4ba4185f4df8113f0c217a9ac8/pkg/fqdn/dnsproxy/proxy.go#L1074

jrajahalme replied:
Added this comment:
```
				// This ReadMsg() will eventually fail due to the read deadline set
				// by 'dns.Client' on the underlying connection when sending the
				// (last) request.
				r, err = c.conn.ReadMsg()
```

bimmlerd replied:
Ah- I was asking about the `ch` read, but while trying to clarify my question I found my mistake- I missed that on getting their response the reading goroutine will send an error to the channels.

bimmlerd reviewed (changes requested):
Requesting changes not because I've found a specific bug, but due to several questions. Nits are nonblocking.

nebril commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -762,13 +762,27 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
 }
 
-func setSoMark(fd int, secId identity.NumericIdentity) error {
+func setSoMarks(fd int, ipFamily ipfamily.IPFamily, secId identity.NumericIdentity) error {
```
Would it be possible to add comments in this function which would explain why we are setting these options? This might be out of scope of this PR, but I cannot figure out why we are setting reuseaddr here based on commit message/code .

jrajahalme replied:
Added :-)

nebril commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -937,38 +1055,66 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
+	key := protocol + "-" + originalSrcAddr.String() + "-" + targetServerAddrStr
+	p.Lock()
+	client := p.DNSClients[key]
+	if client == nil {
```
It would be nice to extract this whole block as `NewDNSClient` function.

joamaki replied:
+1 from me

jrajahalme replied:
Done, also for `GetDNSClient` function, turned out pretty nice :-)

bimmlerd reviewed (changes requested):
Remaining worry to me is the DNSProxy lock per request.

Couldn't find an issue in the intricate locking and cooperative reading of the conn, nice work!

nebril reviewed (approved):
I tried to find holes in the locking logic, looks like you covered all the bases to me.

I added two nitpicks in the comments, feel free to fix them if you end up re-pushing.

joamaki commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -762,16 +768,154 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
+	for {
+		// Try taking the reader lock
+		if c.readerLock.TryLock() {
+			// We are responsible for reading responses for all users
```
This seems unnecessarily complicated and this sort of knife-play with locks is dangerous. I would strongly suggest having a goroutine per client that manages sending the requests on behalf of the users and reading the responses. It'll be much easier to follow.

jrajahalme replied:
Tried it, it is structurally more complex, and adds spawning a new goroutine for each proxied DNS request, something that you have tried to weed off on the agent regeneration side of things.

joamaki replied:
I said goroutine per client above, of course per request would be excessive. E.g. in a really rough sketch:

```
type Request struct {
  Msg *dns.Msg
  ResponseChan <-chan Response
}

type Response struct {
  Msg *dns.Msg
  RoundTripTime time.Duration
  Error error
}

// Dispatcher takes care of the pool of clients and routes it to the right one:

func (d *Dispatcher) Exchange(msg *dns.Msg) (response *dns.Msg) { 
  response := make(chan Response, 1)
  d.requests <- Request{msg, response}
  <-response
}

func (d *Dispatcher) dispatchLoop( ) {
  for {
  select {
  case req := <-d.requests:
     client := d.findOrCreateClient(req.Msg)
     client.requests <- req
  case <-d.Timer.C:
     // cleanup clients 
  case <-d.stop:
     // stop dispatcher
  }
  }
}

// Client takes care of a connection:

func (c *Client) receiveLoop(conn *Conn) {
  for {
     resp, err := c.conn.ReadMsg()
     // send the response to the central event loop
     c.responses <- resp
  }
}

func (c *Client) requestLoop() {
  var requests map[int]Request
  for {
    select {
    case req := <-d.requests:
        requests[req.Msg.Id] = req
        // send the request over the connection
        c.conn.SendContext(req.Msg)

    case resp := <-d.responses:
       if req, ok := requests[resp.Msg.Id]; ok {
         req.response <- resp.Msg
         close(req.response)
       } 

    case <-c.stop:
      // stop the client etc.
    }
}
```

jrajahalme replied:
Clean and nice pattern, thank you.

Due to the restrictions placed on the proxy by in needing to be transparent, i.e., use the original source address and port in the DNS client connection, this pattern leads to, on average, a goroutine per request for dual stack pods, and 2 goroutines per request for single stack  pods. Why is this? Well,

- Since we must use the original source IP address, and we must also intercept the responses, we instruct the networking namespace of the dnsproxy (host) to receive packets destined to a non-local (foreign) IP address via the IP_TRANSPARENT socket option.
- in order to not hijack some random by-standing traffic going to the original pod, we must also use the original port number
- (DNS) clients use ephemeral source ports, i.e., the port can be different in every request. Typically, a DNS resolver library uses the same ephemeral port only to requests from a single "gethostbyname" API call, or equivalent.
- To be able to receive responses to the ephemeral source port, we must have a socket bound to that address:port (for UDP), or a connection from that address:port to the DNS server address:port (for TCP).
- This leads to a new DNS client and socket for every different original source address - ephemeral port pair we see. We also need to make sure these were actually used to communicate with the DNS server, so we use the whole 5-tuple as a key

Why can't we keep DNS clients pooled and ready to receive traffic between client requests then?
- We have no guarantees that the source pod will keep on using the same ephemeral port in future. We've had upstream socket bind errors (in Envoy, where we have operated in this mode for years already) when a client pod has rapidly cycled through its ephemeral port space, e.g. when performing netperf CRR or similar performance tests.
- We could try keep the client and its bound socket around for some minimal time to save resources when a DNS resolver is enumerating through its domain suffix list, where is seems likely that the same source ephemeral port is going to be reused until the resolver gets an actual result with an IPv4/6 address or quits trying. It might be safe to close the client socket only after a response with the `A`/`AAAA` records have been passed back to the pod, or after a timeout of a few milliseconds. This would be something we currently don't do and is prone to socket bind errors, so I'd like to leave this for later.
- So the client socket can not be left lingering around, as it causes network traffic destined for the source pod to be intercepted to the dnsproxy, which is exactly what we want but only until a DNS response has been received.

joamaki replied:
Thanks for the detailed explanation. This is so great I wonder if we should add some of this to code comments?
Given the ephemeral nature and the amount of polish and review, I'm fine with keeping it this way.

pippolo84 replied:
+1 to add this a comment. Thanks for clarifying this!

joamaki commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -913,36 +1066,79 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
-	request.Id = dns.Id() // force a random new ID for this request
-	response, _, err := client.ExchangeWithConn(request, conn)
+	defer func() {
+		p.Lock()
```
The management of the set of clients shouldn't be mixed in here like this.

jrajahalme replied:
Agreed, it is now refactored into `GetDNSClient` function.

joamaki replied:
Awesome thanks!

qmonnet commented on Documentation/contributing/development/debugging.rst:
```diff
@@ -465,7 +465,7 @@ Deadlock detection
 Cilium can be compiled with a build tag ``lockdebug`` which will provide a
 seamless wrapper over the standard mutex types in Golang, via
-`sasha-s/go-deadlock library <https://github.com/sasha-s/go-deadlock>`_. No
+`sasha-s/go-deadlock library <https://github.com/cilium/go-deadlock>`_. No
```
Should this tell that we use a fork? Probably not necessary, but asking just in case.

jrajahalme replied:
This became unnecessary when I moved the shared client code to `cilium/dns` (see https://github.com/cilium/dns/pull/7), where there is no deadlock detection/race builds.

qmonnet reviewed (approved):
Doc change looks good.

pippolo84 reviewed (commented):
Spent quite a bit of time reasoning on the new cooperative logic to share clients and seems sound. :+1:  Also, my understanding is that, with the latest commit, we should pay an overhead due to the new client-sharing logic only for concurrent requests with the same 5-tuple. This should be ok-ish performance wise.

Anyway, I wonder if we should give a try to the model proposed by Jussi [here](https://github.com/cilium/cilium/pull/28928/commits/ef66937d1f865359609faf275cecda1d73eb2186#r1384934221). That's because I'm still a bit concerned that this logic is not super easy to grasp and it fits in a part of the code that is already pretty difficult to follow.

pippolo84 reviewed (approved):
Given the thorough explanation in this [comment](https://github.com/cilium/cilium/pull/28928#discussion_r1385293778), for me it's fine to go ahead with the current client-sharing logic.
Thanks! :100:

jrajahalme committed:
dnsproxy: Use original source address in connections to dns servers

Set transparent, reuseaddr, and reuseport options and use the original
source address on connections from DNS proxy to DNS servers to allow use
of non-local source address as well as recreate sockets on the same
5-tuple without needing to wait for the TCP TIME_WAIT to finish.

Use the MagicMarkEgress mark on connections to the dns servers instead
the generic MagicMarkIdentity.

Use original source address in connections to dns servers when the source
address is not one of the host IPs. The original source address and port
can not be reused if there is already socket with them to the same
destination on the same networking namespace.

Use new dns.SharedClients to reuse DNS clients between all requests that
originate from the same source address and port. This allows multiple
different requests to be pending at the same time on the same dns Client,
which happens whenever the source pod sends multiple DNS requests from
the same resolver invocation, e.g., for A and AAAA records.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jschwinger233 reviewed (approved):
Thanks Jarno! Just tested locally and it works perfectly, even forward packets' marks have been fixed!

jschwinger233 reviewed (commented):
One more thing, do you think we should submit changes under vendor/ to github.com/cilium/dns?

bimmlerd reviewed (approved):
Couldn't solidify my hunch of a performance regression and don't have a better design to propose, no reason to hold this back further. Thanks for addressing all the feedback, Jarno!
